src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+    VerifyOrReturnError(IsFabricGroupId(mapping.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+    FabricData fabric(fabric_index);
+
+    if (CHIP_NO_ERROR != fabric.Load(mStorage) || kInvalidEndpointId == fabric.first_endpoint)
+    {
+        // New fabric, or new endpoint
+        GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name).Save(mStorage);
+        EndpointData(fabric_index, mapping.endpoint, mapping.group).Save(mStorage);
+        fabric.first_endpoint = mapping.endpoint;
+        fabric.endpoint_count = 1;
+        return fabric.Save(mStorage);
+    }
+
+    // Existing fabric
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    // Loop through the fabric's endpoints
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            // Endpoint info not found
+            break;
+        }
+        if (endpoint.id == mapping.endpoint)
+        {
+            // Target endpoint info found
+            break;
+        }
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    if (endpoint.id != mapping.endpoint)
+    {
+        // Endpoint not found, create new, and insert first
+        GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name).Save(mStorage);
+        EndpointData(fabric_index, mapping.endpoint, mapping.group, fabric.first_endpoint).Save(mStorage);
+        fabric.first_endpoint = mapping.endpoint;
+        fabric.endpoint_count++;
+        return fabric.Save(mStorage);
+    }
+
+    // Endpoint info found, loop endpoints
+
+    GroupData group(fabric_index, endpoint.id, endpoint.first_group);
+
+    do
+    {
+        if (CHIP_NO_ERROR != group.Load(mStorage))
+        {
+            break;
+        }
+        if (group.group == mapping.group)
+        {
+            // Duplicated group
+            return CHIP_NO_ERROR;
+        }
+        group.group = group.next;
+    } while (kUndefinedGroupId != group.group);
+
+    // Insert group first, update endpoint
+    GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name, endpoint.first_group).Save(mStorage);
+    endpoint.first_group = mapping.group;
+    return endpoint.Save(mStorage);
+}
+
+CHIP_ERROR GroupDataProviderImpl::RemoveGroupMapping(chip::FabricIndex fabric_index, const GroupMapping & mapping)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(kInvalidEndpointId != mapping.endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(IsFabricGroupId(mapping.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+    FabricData fabric(fabric_index);
+
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_INVALID_FABRIC_ID);
+    VerifyOrReturnError(kInvalidEndpointId != fabric.first_endpoint, CHIP_ERROR_KEY_NOT_FOUND);
```

Comment:
Shouldn't the above have caused CHIP_ERROR_KEY_NOT_FOUND from Load? What do we do if this condition occurs here? It means we loaded an invalid record at the key, We should delete it since it's incoherent.

Reply:
Not really. The fabric structures hold the root of both the group mappings, and the keysets, which are orthogonal. The fabric info may exist, and point to existing keysets, and yet no group mappings may exist.